core-backend/src/accessors.rs

```diff
@@ -0,0 +1,525 @@
+        ctx: &mut MemoryCallerContext<Caller>,
+    ) -> Self
+    where
+        Caller: AsContextExt<State = HostState<Ext, BackendMemory<ExecutorMemory>>>,
+        Ext: BackendExternalities + 'static,
+    {
+        Self {
+            result: ctx
+                .memory_wrap
+                .io_mut_ref()
+                .and_then(|io| io.read_as(&mut ctx.caller_wrap, output)),
+        }
+    }
+}
+
+impl<T: Decode + MaxEncodedLen> SyscallArg for ReadDecoded<T> {
+    type Output = WasmMemoryReadDecoded<T>;
+    const REQUIRED_ARGS: usize = 1;
+
+    fn pre_process<Caller, Ext>(
+        registry: &mut Option<MemoryAccessRegistry<Caller>>,
+        args: &[Value],
+    ) -> Result<Self::Output, HostError>
+    where
+        Caller: AsContextExt<State = HostState<Ext, BackendMemory<ExecutorMemory>>>,
+        Ext: BackendExternalities + 'static,
+    {
+        debug_assert_eq!(args.len(), Self::REQUIRED_ARGS);
+
+        let ptr = SyscallValue(args[0]).try_into()?;
+
+        Ok(registry.get_or_insert_default().register_read_decoded(ptr))
+    }
+
+    fn post_process<Caller, Ext>(
+        output: Self::Output,
+        ctx: &mut MemoryCallerContext<Caller>,
+    ) -> Self
+    where
+        Caller: AsContextExt<State = HostState<Ext, BackendMemory<ExecutorMemory>>>,
+        Ext: BackendExternalities + 'static,
+    {
+        Self {
+            result: ctx
+                .memory_wrap
+                .io_mut_ref()
+                .and_then(|io| io.read_decoded(&mut ctx.caller_wrap, output)),
+        }
+    }
+}
+
+impl<T: Decode + MaxEncodedLen + Default> SyscallArg for ReadDecodedSpecial<T> {
+    type Output = Option<WasmMemoryReadDecoded<T>>;
+    const REQUIRED_ARGS: usize = 1;
+
+    fn pre_process<Caller, Ext>(
+        registry: &mut Option<MemoryAccessRegistry<Caller>>,
+        args: &[Value],
+    ) -> Result<Self::Output, HostError>
+    where
+        Caller: AsContextExt<State = HostState<Ext, BackendMemory<ExecutorMemory>>>,
+        Ext: BackendExternalities + 'static,
+    {
+        debug_assert_eq!(args.len(), Self::REQUIRED_ARGS);
+
+        let ptr = SyscallValue(args[0]).try_into()?;
+
+        if ptr != PTR_SPECIAL {
+            Ok(Some(
+                registry.get_or_insert_default().register_read_decoded(ptr),
+            ))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn post_process<Caller, Ext>(
+        output: Self::Output,
+        ctx: &mut MemoryCallerContext<Caller>,
+    ) -> Self
+    where
+        Caller: AsContextExt<State = HostState<Ext, BackendMemory<ExecutorMemory>>>,
+        Ext: BackendExternalities + 'static,
+    {
+        match output {
+            Some(output) => Self {
+                result: ctx
+                    .memory_wrap
+                    .io_mut_ref()
+                    .and_then(|io| io.read_decoded(&mut ctx.caller_wrap, output)),
+            },
+            None => Self {
+                result: Ok(Default::default()),
```

Comment:
should we return option for the case?

Reply:
The original code from `fn register_and_read_value` returned 0 if the address matched `PTR_SPECIAL`, I think this is the desired behavior...

Reply:
just from coding perspective 
meet PTR_SPECIAL => return None => unwrap_or_default 
but not a big deal actually